src/servers/src/grpc/handler.rs

```diff
@@ -145,25 +145,35 @@ impl GreptimeRequestHandler {
 }
 
 pub(crate) fn create_query_context(header: Option<&RequestHeader>) -> QueryContextRef {
-    let ctx = QueryContext::arc();
-    if let Some(header) = header {
-        // We provide dbname field in newer versions of protos/sdks
-        // parse dbname from header in priority
-        if !header.dbname.is_empty() {
-            let (catalog, schema) =
-                crate::parse_catalog_and_schema_from_client_database_name(&header.dbname);
-            ctx.set_current_catalog(catalog);
-            ctx.set_current_schema(schema);
-        } else {
-            if !header.catalog.is_empty() {
-                ctx.set_current_catalog(&header.catalog);
-            }
-            if !header.schema.is_empty() {
-                ctx.set_current_schema(&header.schema);
+    let (catalog, schema) = header
+        .map(|header| {
+            // We provide dbname field in newer versions of protos/sdks
+            // parse dbname from header in priority
+            if !header.dbname.is_empty() {
+                crate::parse_catalog_and_schema_from_client_database_name(&header.dbname)
+            } else {
+                (
+                    if !header.catalog.is_empty() {
+                        &header.catalog
+                    } else {
+                        DEFAULT_CATALOG_NAME
+                    },
+                    if !header.schema.is_empty() {
+                        &header.schema
+                    } else {
+                        DEFAULT_SCHEMA_NAME
+                    },
+                )
             }
-        }
-    };
-    ctx
+        })
+        .unwrap_or((DEFAULT_CATALOG_NAME, DEFAULT_SCHEMA_NAME));
+
+    QueryContextBuilder::new()
+        .catalog(catalog.to_string())
+        .schema(schema.to_string())
```

Comment:
By the way, perhaps we can make catalog/schema/traceid read-only fields with the new design of `QueryContext`.

Reply:
Good point. I'd like to refactor QueryContext related code in a separate pr since I believe the change would be massive  😆